### DISCUSSION TITLE
Room, Vote, JobTarget, Chat 분리에 대해서

### DISCUSSION THREAD

**Discussion by waterricecake** [mafia-together/backend#80] [General]
Redis 도입전 해당 부분들이 분리될 필요가 없기에 같이 구현하였지만 Redis를 도입하면서 든 생각이 Chat은 후에 Pub/Sub 구조를 가져가게 될 것인데 분리를 해야하지 않을까 라는 생각을 가지게 되었습니다.

왜냐하면 PUB/SUB은 저장한 채팅들을 Queue형식으로 빼주는 방법으로 구현되기 때문입니다. 이때문에 Room을 그냥 RedisHash로 저장하는 방식은 어울리지 않다고 생각하였습니다.

이때문에 Chat을 분리하는 게 좋겠다고 생각했는데 하다보니 Vote, JobTarget또한 분리되는게 좋겠다고 생각이 들더라고요.
그이유는 Room과 생명주기가 다르기 때문이라고 생각합니다.

Vote와 JobTarget은 각각 투표, 밤 상태가 종료시 초기화 되는데 이 초기화가 어떻게 보면 이 Vote, JobTarget의 생명주기가 됩니다. 그런데 만약 Room에 같이 있을 경우 플레이어 하나 바뀔때마다 새로 room 전체를 다시 넣어주고 하는 작업이 들더라고요. 그것보다는 Vote, JobTarget을 분리하여 관리하는 것이 더 자연스럽다고 생각하게 되었습니다.

이러한 방법을 채택한다면 프로젝트 초기 제가 고민하였던 RoomManager의 의존성 문제도 해결할 수 있을 것이라 생각합니다.

**Comment by kpeel5839** [comment]
애그리거트 단위로 나누는 것은 무조건 필요하다고 생각했어요.

최근들어 제가 이벤트 스토밍을 진행해보았는데요. 해당 기법이 도메인에 대한 이해도도 높일 수 있고, 도메인을 애그리거트 단위로 나누는데 도움을 많이 주는 것 같더라고요.

그래서 건의 드려봅니다. 제가 아직까지는 여러분들보다 도메인에 대한 이해도도 떨어지고 이참에 애그리거트 단위로 도메인을 분리하고 서로 간의 의존성도 완전히 정리할겸 이벤트 스토밍을 한번 해보는 것 어떨까요?

부정적인 피드백도 괜찮아요! 아무래도 이미 도메인이 구성되어 있는 상태에서 이벤트 스토밍을 한다는 것 자체가 과한 비용을 투자하는 것일수도 있으니까요! 차라리 만나서 컴팩트하게 1~2시간 정도, 현재 도메인을 애그리거트 단위로 분리하는 것 정도로 충분할 것 같기도 하고요.

**Reply by waterricecake** [reply]
저도 이번에 Redis 적용해보면서 가장 많이 건들여야겠는 부분이 이벤트더라고요 근데 문제가 한 도메인에 여러가지 애그리거트가 묶여있는 상황인거 같아서 RegisterEvent라는 엔티티 내부에서 이벤트 발행하는 매서드로 억지로 발행하는 느낌이 계속 들더라고요. 그래서 저도 차라리 한번 만나서 도매인 애그리거트 단위 분리를 한번 해보는게 좋을것 같습니다. @thdwoqor

**Reply by kpeel5839** [reply]
네네! 좋아요!

근데 제가 이벤트 스토밍에 대한 설명이 부족했던 것 같아요.

![image](https://github.com/mafia-together/backend/assets/89840550/77642fd8-f39a-4bbc-8788-0e96a701e357)
> 진행하게 된다면, 세부적인 내용은 제가 안내해드릴 수 있을 것 같아서 이미지는 대충 느낌으로만 가져왔어요

이벤트 스토밍은 저런 형식으로 정말 이벤트들을 기반으로 브레인 스토밍하는 것을 의미해요.
그래서 저희 서비스에서 어떠한 이벤트가 발생하는지, 그 이벤트에 대한 트리거, 액터는 누구이고 해당 이벤트는 어떠한 애그리거트에 속하게 되는지

또한 도메인 정책상 어떠한 애그리거트에서 다른 애그리거트를 변화시키는 일이 있는지, 어떠한 외부시스템을 사용하는지, 이런 것들을 한눈에 볼 수 있도록 정리할 수 있거든요.
그래서 건의 드려봤어요.

제가 진행하고 있는 다른 프로젝트에서는 이벤트 스토밍의 결과로 다음과 같은 구조가 나오게 되었어요. (아직 다하지는 않았어요.)
![image](https://github.com/mafia-together/backend/assets/89840550/51a16bf0-0c55-4067-857d-93faa7f07f76)

**Reply by thdwoqor** [reply]
지금 룸에 모든게 다들어가있어서 분리하면 괜찮을거같아요
이벤트 스토밍은 새로 오신 프론트 분들이 도메인 이해를 높이는 데 도움이 될 것 같아요. 
기존에 있던사람도 서로 잘못 이해하고 있던 부분을 정리하는 데 도움이 될 같구요

**Reply by kpeel5839** [reply]
오호! 재백짱은 동의하는군요

**Reply by waterricecake** [reply]
아 그이벤트가 그 이벤트가 아니구나 ㅋㅋㅋㅋㅋㅋ 그러면 다음 개발자 회의 전에 한번 모여서 정리해보죠! 언제가 좋슴까?
아니면 재백짱의 말은 이벤트 스토밍을 개발자 회의때 프론트분들과 함께 하자는 의미인가요?

**Reply by kpeel5839** [reply]
저는 괜찮은 것 같아요. 조금 오래 걸리는 작업이기는한데

**Reply by thdwoqor** [reply]
ㅇㅇ 우리보다 새로들어오신분들한테 더 도움이 많이될꺼같아서
